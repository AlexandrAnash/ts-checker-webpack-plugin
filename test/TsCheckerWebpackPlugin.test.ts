import * as path from "path";
import * as fs from "fs";
import webpack = require("webpack");
import MemoryFs = require("memory-fs");

const testCasesPath = path.join(__dirname, "testCases");
const tests = fs.readdirSync(testCasesPath).filter(dir => fs.statSync(path.join(testCasesPath, dir)).isDirectory());

const basePath = path.resolve(__dirname, "../");
const normalizeError = (error: string) =>
  error.replace(/\r\n?/g, "\n").replace(new RegExp(path.join(basePath), "g"), "Xdir");

/**
 * Dummy test
 */
describe("TsCheckerWebpackPlugin", () => {
  tests.forEach(testName => {
    const testPath = path.join(testCasesPath, testName);
    const webpackConfigPath = path.join(testPath, "webpack.config.ts");
    const webpackConfig = require(webpackConfigPath);

    it(testName, done => {
      const c = webpack(webpackConfig);
      const compilers = (c as any).compilers ? (c as any).compilers : [c];

      compilers.forEach((compiler: any) => {
        const ifs = compiler.inputFileSystem;
        compiler.inputFileSystem = Object.create(ifs);
        compiler.outputFileSystem = new MemoryFs();
        compiler.inputFileSystem.readFile = function() {
          const args = Array.prototype.slice.call(arguments);
          const callback = args.pop();
          ifs.readFile.apply(
            ifs,
            args.concat([
              (err: any, result: any) => {
                if (err) return callback(err);
                callback(null, result.toString("utf-8").replace(/\r/g, ""));
              },
            ])
          );
        };
        compiler.apply(new webpack.optimize.OccurrenceOrderPlugin(false));
      });

      c.run((err: Error, stats: webpack.Stats) => {
        expect(err).toBeFalsy();

        const { errors, warnings } = stats.toJson();

        if (stats.hasErrors()) {
          const normalizedErrors = errors.map(normalizeError);
          expect(normalizedErrors).toMatchSnapshot();
        }

        if (stats.hasWarnings()) {
          const normalizedWarnings = warnings.map(normalizeError);
          expect(normalizedWarnings).toMatchSnapshot();
        }

        done();
      });
    });
  });
});
