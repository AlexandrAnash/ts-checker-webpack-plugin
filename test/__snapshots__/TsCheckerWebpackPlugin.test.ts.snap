// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsCheckerWebpackPlugin basic-type-error 1`] = `
Array [
  "Xdir/test/testCases/basic-type-error/src/entry.ts
Type '\\"test\\"' is not assignable to type 'number'.",
]
`;

exports[`TsCheckerWebpackPlugin ts-dependency-only-error 1`] = `
Array [
  "Xdir/test/testCases/ts-dependency-only-error/src/entry.ts
Class 'TestImpl' incorrectly implements interface 'Test'.
  Property 'doStuff' is missing in type 'TestImpl'.",
]
`;

exports[`TsCheckerWebpackPlugin use-library-type-definition-error 1`] = `
Array [
  "Xdir/test/testCases/use-library-type-definition-error/src/entry.ts
Could not find a declaration file for module 'react'. 'Xdir/node_modules/react/react.js' implicitly has an 'any' type.",
]
`;
